dot/rpc/modules/state.go

```diff
@@ -489,6 +495,41 @@ func (sm *StateModule) QueryStorage(
 	return nil
 }
 
+// QueryStorageAt queries historical storage entries (by key) at the best block if the given block hash is nil
+func (sm *StateModule) QueryStorageAt(
+	_ *http.Request, req *StateStorageQueryAtRequest, res *[]StorageChangeSetResponse) error {
+	var atBlockHash common.Hash
+	if req.At.IsEmpty() {
+		atBlockHash = sm.blockAPI.BestBlockHash()
+	} else {
+		atBlockHash = req.At
+	}
+
+	response := make([]StorageChangeSetResponse, 0, 1)
+	changes := make([][2]*string, 0, len(req.Keys))
+
+	for _, key := range req.Keys {
+		value, err := sm.storageAPI.GetStorageByBlockHash(&atBlockHash, common.MustHexToBytes(key))
+		if err != nil {
+			return fmt.Errorf("getting value by block hash: %w", err)
+		}
+		var hexValue *string
+		if len(value) > 0 {
```

Comment:
should we differentiate `[]byte{}` and `[]byte(nil)`?? This comes from the trie and the empty byte slice means it's set. The `nil` byte slice means the value doesn't exist though. I would tend to think we should return `0x` in case the value returned is an empty-non-nil byte slice.

Reply:
Good point, I've updated this function and `state_QueryStorage` to handle `[]byte{}` and `[]byte(nil)` and added tests for both conditions.